content/en/blog/_posts/2024-04-01-validating-admission-policy-ga/index.md

```diff
@@ -0,0 +1,216 @@
+    message: 'all containers must NOT set privileged to true'
+```
+Check its status again, and you should see all warnings cleared.
+
+Next, let's create a namespace for our tests.
+```shell
+kubectl create namespace policy-test
+```
+Then, bind the policy to the namespace. But at this point, we set the action to `Warn` 
+so that the policy prints out warnings instead of rejecting the requests.
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicyBinding
+metadata:
+  name: "pod-security.policy-binding.example.com"
+spec:
+  policyName: "pod-security.policy.example.com"
+  validationActions: ["Warn"]
+  matchResources:
+    namespaceSelector:
+      matchLabels:
+        "kubernetes.io/metadata.name": "policy-test"
+```
+Tests out policy enforcement.
+```shell
+kubectl create -n policy-test -f- <<EOF
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    app: nginx
+  name: nginx
+spec:
+  selector:
+    matchLabels:
+      app: nginx
+  template:
+    metadata:
+      labels:
+        app: nginx
+    spec:
+      containers:
+      - image: nginx
+        name: nginx
+        securityContext:
+          privileged: true
+          allowPrivilegeEscalation: true
+EOF
+```
+```text
+Warning: Validation failed for ValidatingAdmissionPolicy 'pod-security.policy.example.com' with binding 'pod-security.policy-binding.example.com': all containers must set runAsNonRoot to true
+Warning: Validation failed for ValidatingAdmissionPolicy 'pod-security.policy.example.com' with binding 'pod-security.policy-binding.example.com': all containers must set readOnlyRootFilesystem to true
+Warning: Validation failed for ValidatingAdmissionPolicy 'pod-security.policy.example.com' with binding 'pod-security.policy-binding.example.com': all containers must NOT set allowPrivilegeEscalation to true
+Warning: Validation failed for ValidatingAdmissionPolicy 'pod-security.policy.example.com' with binding 'pod-security.policy-binding.example.com': all containers must NOT set privileged to true
+Error from server: error when creating "STDIN": admission webhook "webhook.example.com" denied the request: [container "nginx" must set RunAsNonRoot to true in its SecurityContext, container "nginx" must set ReadOnlyRootFilesystem to true in its SecurityContext, container "nginx" must NOT set AllowPrivilegeEscalation to true in its SecurityContext, container "nginx" must NOT set Privileged to true in its SecurityContext]
+```
+Looks great! The policy and the webhook give equivalent results.
+After a few other cases, when we are confident with our policy, maybe it is time to do some cleanup.
+
+- For every expression, we repeat access to `object.spec.template.spec.containers` and each `securityContext` in every expression;
+- There is a pattern of checking presence of a field and then accessing it, which looks a bit verbose.
+
+Fortunately, with the release of Kubernetes 1.28, we have new solutions for both issues.
```

Comment:
Is this right? v1.28 is not new.

Reply:
Yes it is correct. I've created an 1.28 cluster to validate.